File: examples/dda/level-dda.lua
return function(input)
    local behavior = (input.state and input.state.behavior) or {}
    local recent = behavior.recent or {}
    local recentSteps = behavior.recentSteps or {}
    local config = input.config or {}
    local weights = config.behaviorWeights or {
        level_failed = 1,
        level_retry = 0.75,
        item_used = 0.25,
        mistake = 0.5,
    }

    local friction = 0
    for behaviorType, weight in pairs(weights) do
        friction = friction + (tonumber(recent[behaviorType]) or 0) * (tonumber(weight) or 0)
    end
    local recentStepCount = #recentSteps
    local frictionPerStep = friction / math.max(1, recentStepCount)
    local minRecentSteps = tonumber(config.minRecentSteps) or 3
    local decreaseThreshold = tonumber(config.decreaseThreshold) or 0.8
    local increaseThreshold = tonumber(config.increaseThreshold) or 0

    local adjustment = "keep"
    if recentStepCount >= minRecentSteps then
        if frictionPerStep >= decreaseThreshold then
            adjustment = "decrease"
        elseif frictionPerStep <= increaseThreshold then
            adjustment = "increase"
        end
    end

    return {
        payload = { adjustment = adjustment },
        diagnostics = {
            friction = friction,
            frictionPerStep = frictionPerStep,
            recentStepCount = recentStepCount,
        },
    }
end

File: examples/dda/level-dda.js
function execute(input) {
  const behavior = input.state?.behavior ?? {};
  const recent = behavior.recent ?? {};
  const recentSteps = behavior.recentSteps ?? [];
  const config = input.config ?? {};
  const weights = config.behaviorWeights ?? {
    level_failed: 1,
    level_retry: 0.75,
    item_used: 0.25,
    mistake: 0.5,
  };

  let friction = 0;
  for (const [type, weight] of Object.entries(weights)) {
    friction += Number(recent[type] ?? 0) * Number(weight ?? 0);
  }
  const frictionPerStep = friction / Math.max(1, recentSteps.length);
  const minRecentSteps = Number(config.minRecentSteps ?? 3);
  const decreaseThreshold = Number(config.decreaseThreshold ?? 0.8);
  const increaseThreshold = Number(config.increaseThreshold ?? 0);

  let adjustment = "keep";
  if (recentSteps.length >= minRecentSteps) {
    if (frictionPerStep >= decreaseThreshold) adjustment = "decrease";
    else if (frictionPerStep <= increaseThreshold) adjustment = "increase";
  }

  return {
    payload: { adjustment },
    diagnostics: {
      friction,
      frictionPerStep,
      recentStepCount: recentSteps.length,
    },
  };
}
